packages/gestalt/src/Masonry.tsx

```diff
@@ -701,15 +709,29 @@ export default class Masonry<T> extends ReactComponent<Props<T>, State<T>> {
       // Full layout is possible
       const itemsToRender = items.filter((item) => item && measurementStore.has(item));
       const itemsWithoutPositions = items.filter((item) => item && !positionStore.has(item));
-      const hasMultiColumnItems =
+      const nextMultiColumnItem =
         _getColumnSpanConfig &&
-        itemsWithoutPositions.some((item) => _getColumnSpanConfig(item) !== 1);
+        itemsWithoutPositions.find((item) => _getColumnSpanConfig(item) !== 1);
 
-      // If there are 2-col items, we need to measure more items to ensure we have enough possible layouts to find a suitable one
-      // we need the batch size (number of one column items for the graph) + 1 (two column item)
-      const itemsToMeasureCount = hasMultiColumnItems
-        ? MULTI_COL_ITEMS_MEASURE_BATCH_SIZE + 1
-        : minCols;
+      let batchSize;
+      if (nextMultiColumnItem) {
+        const gridSize = getColumnCount({ gutter, columnWidth, width, minCols, layout });
+
+        const moduleSize = calculateActualColumnSpan({
```

Comment:
This was the tricky part of this implementation. These two values are easy to find in `multiColumnLayout.ts`, but are not ready at this scope.

- For grid size, I created a whole new util that abstract the way we calculate the column count from both `DefaultLayout` and `FullWidthLayout` (updating those utils to avoid code duplicity too).
- For moduleSize, I exposed the `calculateActualColumnSpan` from  `multiColumnLayout.ts` to be able to use it here.

All of this happens also in MasonryV2.